### PR TITLE
Fix eigen quaternion coeffs problems

### PR DESCRIPTION
**Pull request by KipHamiltons**
Eigen::Quaternion `.coeffs()` returns the coefficients in a different order to the order that the constructor expects.
This PR looks to find and fix all the places where I screwed that up.

### CODE REVIEW

File: module/motion/QuinticWalk/src/QuinticWalk.cpp [hunk 1]
```diff
@@ -219,22 +219,20 @@ namespace module::motion {
...
-            const float halfYaw   = yaw * 0.5f;
-            const float halfPitch = pitch * 0.5f;
-            const float halfRoll  = roll * 0.5f;
-            const float cosYaw    = std::cos(halfYaw);
-            const float sinYaw    = std::sin(halfYaw);
-            const float cosPitch  = std::cos(halfPitch);
-            const float sinPitch  = std::sin(halfPitch);
-            const float cosRoll   = std::cos(halfRoll);
-            const float sinRoll   = std::sin(halfRoll);
-            return Eigen::Quaternionf(cosRoll * cosPitch * cosYaw + sinRoll * sinPitch * sinYaw,  // formerly yzx
```
Comment (TrentHouliston) [inline module/motion/QuinticWalk/src/QuinticWalk.cpp:231]: I'm not actually sure I like this change, it makes things more complicated...
Maybe we just need to document what's going on whenever we use the various constructors.
Or at the very least just wrap it so we still return directly but have the vector4f in the quaternion constructor

File: module/input/SensorFilter/src/MotionModel.hpp [hunk 2]
```diff
@@ -158,22 +158,15 @@ namespace module::input {
...
-            const Eigen::Quaternion<Scalar> dq_dt = Eigen::Quaternion<Scalar>(0.0,
-                                                                              newState.omegaTTt.x() * 0.5,
-                                                                              newState.omegaTTt.y() * 0.5,
-                                                                              newState.omegaTTt.z() * 0.5)
-                                                    * Rwt;
+            const Eigen::Matrix<Scalar, 4, 1> half_omegaTTt_coeffs(newState.omegaTTt.x() * 0.5,
+                                                                   newState.omegaTTt.y() * 0.5,
+                                                                   newState.omegaTTt.z() * 0.5,
+                                                                   0.0);
+            const Eigen::Quaternion<Scalar> dq_dt = Eigen::Quaternion<Scalar>(half_omegaTTt_coeffs) * Rwt;
```
Comment (TrentHouliston) [inline module/input/SensorFilter/src/MotionModel.hpp:165]: I would just wrap these straight into the constructor, e.g. `Quat(Vec(a,b,c,d));`

**Review by TrentHouliston — CHANGES_REQUESTED** [review]

**Comment by KipHamiltons** [comment]
I agree. Both of those are definitely nicer wrapped in a temporary vector. I also noticed another simplification I could make, which I added in a follow up commit.

File: module/motion/QuinticWalk/src/QuinticWalk.cpp (hunk 1 above)
Comment (Bidski) [reply module/motion/QuinticWalk/src/QuinticWalk.cpp:231]: Goddamn this is horrible :vomiting_face:

**Review by Bidski — DISMISSED** [review]
I dont see anything obviously wrong

**Comment by KipHamiltons** [comment]
@TrentHouliston can I grab a review pls :pleading_face:

File: module/input/SensorFilter/src/MotionModel.hpp [hunk 3]
```diff
@@ -206,7 +207,7 @@ namespace module::input {
 
         Eigen::Matrix<Scalar, 4, 1> predict(const Eigen::Matrix<Scalar, size, 1>& state,
                                             const MeasurementType::FLAT_FOOT_ORIENTATION&) {
-            return StateVec(state).Rwt.coeffs();
+            return Eigen::Matrix<Scalar, 4, 1>(state.template segment<4>(StateVec::QX));
```
Comment (TrentHouliston) [inline module/input/SensorFilter/src/MotionModel.hpp:210]: Can't you just do this?
```suggestion
            return state.template segment<4>(StateVec::QX);
```

**Review by TrentHouliston — CHANGES_REQUESTED** [review]

File: module/input/SensorFilter/src/MotionModel.hpp (hunk 3 above)
Comment (KipHamiltons) [reply module/input/SensorFilter/src/MotionModel.hpp:210]: Yep. Good eye.

**Review by Bidski — APPROVED** [review]
I guess?

**Review by Bidski — DISMISSED** [review]
LGTM

File: module/input/SensorFilter/src/MotionModel.hpp [hunk 4]
```diff
@@ -158,15 +158,16 @@ namespace module::input {
...
+                Eigen::Quaternion<Scalar>(Eigen::Matrix<Scalar, 4, 1>(newState.omegaTTt.x() * 0.5,
+                                                                      newState.omegaTTt.y() * 0.5,
+                                                                      newState.omegaTTt.z() * 0.5,
+                                                                      0.0))
+                * Rwt;
             // The change in the rotation is the derivative times the differential (which is the time-step)
-            const Eigen::Quaternion<Scalar> change = Eigen::Quaternion<Scalar>(deltaT * dq_dt.coeffs());
-            // We can add the change to the original, as long as our time step is small enough
-            newState.Rwt = Eigen::Quaternion<Scalar>(Rwt.coeffs() + change.coeffs()).normalized();
+            const auto change = deltaT * dq_dt.coeffs();
```
Comment (TrentHouliston) [inline module/input/SensorFilter/src/MotionModel.hpp:168]: Do not use auto when you are working with Eigen math operations.
https://eigen.tuxfamily.org/dox/TopicPitfalls.html
You fell for one of the classic blunders

**Review by TrentHouliston — CHANGES_REQUESTED** [review]
Fix the bad thing, rest seems ok

File: module/input/SensorFilter/src/MotionModel.hpp (hunk 4 above)
Comment (tayloryoung6396) [reply module/input/SensorFilter/src/MotionModel.hpp:168]:
``` suggestion 
const Eigen::Matrix<double, 4, 1> change(deltaT * dq_dt.coeffs());
```

Comment (KipHamiltons) [reply module/input/SensorFilter/src/MotionModel.hpp:168]: Imma have to repost the meme :joy:

File: module/input/SensorFilter/src/MotionModel.hpp [hunk 5]
```diff
@@ -158,15 +158,16 @@ namespace module::input {
...
+                Eigen::Quaternion<Scalar>(Eigen::Matrix<Scalar, 4, 1>(newState.omegaTTt.x() * 0.5,
+                                                                      newState.omegaTTt.y() * 0.5,
+                                                                      newState.omegaTTt.z() * 0.5,
+                                                                      0.0))
+                * Rwt;
             // The change in the rotation is the derivative times the differential (which is the time-step)
-            const Eigen::Quaternion<Scalar> change = Eigen::Quaternion<Scalar>(deltaT * dq_dt.coeffs());
-            // We can add the change to the original, as long as our time step is small enough
-            newState.Rwt = Eigen::Quaternion<Scalar>(Rwt.coeffs() + change.coeffs()).normalized();
+const Eigen::Matrix<double, 4, 1> change(deltaT * dq_dt.coeffs());
```
Comment (KipHamiltons) [inline module/input/SensorFilter/src/MotionModel.hpp:168]:
```suggestion
const Eigen::Matrix<Scalar, 4, 1> change(deltaT * dq_dt.coeffs());
```

File: module/input/SensorFilter/src/MotionModel.hpp (hunk 4 above)
Comment (KipHamiltons) [reply module/input/SensorFilter/src/MotionModel.hpp]: Well that page is quite interesting. Especially the part about passing by value. that makes things a bit spicy

**Review by TrentHouliston — APPROVED** [review]
Sure

**Pull request merged**